### ISSUE TITLE
API mismatch with selection ranges

### ISSUE THREAD

**Issue by dbalatero** [asmagill/hs._asm.axuielement#17]
Hey @asmagill! Just wanted to point out a small inconsistency in API with this library:

Calling `element:attributeValue("AXSelectedTextRange")` returns the following data structure:

```lua
{
  loc = 50,
  len = 2
}
```

However, when you pass in a data structure that you want converted to a `UITextRange`, you have to pass it in like this:

```lua
element:setSelectedTextRange({
  location = 50,
  length = 2
})
```

I spent a while beating my head against the wall trying to figure out why setting the text range didn't work until I dug into the internals, which is probably not the ideal experience!

I think it would make sense to mirror the data structure that OS X provides. At this point for backwards compatibility you'd probably have to support both.

Basically this code would need to change, from my read:

https://github.com/asmagill/hs._asm.axuielement/blob/575b0df/common.m#L295-L297

If you're interested I can try to make a patch to support either object style?

**Comment by asmagill** [comment]
Agreed, they should be the same... I'll have to look and see which is easier.

Since we're looking towards moving this to core soon, how big of a deal will breaking changes be to fix in your code? I'd rather a standard clean interface then a bunch of extra logic to force backwards compatibility for something that will actually be "new" to most users.

**Comment by dbalatero** [comment]
@asmagill I assume the namespace is going to slightly change when you merge into core, right? Dropping the `_asm` prefix? I vendor axuielement in my project anyways so I can upgrade at my leisure.

**Comment by asmagill** [comment]
At a minimum, yes, the name will change, but in most cases that just means changing the initial `require` line.

But I'm not ruling out other potential changes (like standardizing on loc/len or location/length) and maybe even changes to the metamethod shortcuts... probably not to the "official" methods names, though.

Basically, I was trying a few different syntaxes to see what seemed the most "natural", but if you look at the module's `init.lua` code, you'll see that there is a lot of logic there that could probably be simplified or streamlined... not sure yet.

There are also a couple of potentially long running sequences, especially with capturing all children objects and filters with queries, that I'd like to maybe make more responsive by incorporating the coroutine support that https://github.com/Hammerspoon/hammerspoon/pull/2308 would give us, but I can probably phase that in after the pull lands, since that's going to require a lot of testing before I think it will land.

Basically, I'm trying to get a feel for how big of a deal backwards compatibility is (not that I have specific breaking changes in mind) since this isn't in core yet... I'm a big fan of backwards compatibility for things in the official core, but, for the moment, it can be argued that this isn't "official" yet.

That said, you guys provide a lot of support and ideas and features, so I don't want to alienate you either 😜

**Comment by dbalatero** [comment]
@asmagill I think as long as you can remember to keep a running changelog of what APIs you break, then I can grep my codebase and conform to the new standards.

I can also stay on your old version while I do that work, so it's not necessarily that you're going to break me the minute you release it to core either!

**Comment by asmagill** [comment]
The naming mismatch for ranges has been fixed.

If you're still running an earlier version of the module, you should probably get ready updating soon.

The current API can be read at https://github.com/asmagill/hs._asm.axuielement/blob/master/Reference_Core.md but the summary can be thought of as:

* you can no longer to attribute names *without* the AX prefix when using the metamethod shortcuts (e.g. `obj.AXTitle` instead of `obj.title` or `obj.Title` -- this was done to simplify the internal logic and to remove the confusion when trying to remember that `title`, `Title`, and `AXTitle` might actually be referring to the same thing.
* the `__call` metamethod has been removed, so anything of the form `obj(...)` where `obj` is an `hs.axuielement` will have to be rewritten as one of the following (depending upon how it was used):
  * `obj.AXAttribute` instead of `obj("AXAttribute")`
  * `obj.AXAttribute = value` instead of `obj("setAXAttribute", value)`
  * `obj:doAXAction()` instead of `obj("doAXAction")`
  * `obj:AXParameterizedAttributeWithParamter(param)` instead of `obj("AXParameterizedAttributeWithParameter", param)`

I don't know if you used these (they were poorly documented before, and I know that most people didn't use them, but just in case):
* `hs.axuielement:matchesCriteria` has been rewritten and the syntax changed. The changes *also* allow for:
  * matching whether attribute(s) exist and have *any* value (except nil)
  * matching whether attribute(s) exist and have *no* assigned value (i.e. nil)
  * matching whether attribute(s) exist (and we don't care if they have an assigned value or not)
  * inverting the match result (i.e. return true if the criteria does *not* match for the element)
* `hs.axuielement:elementSearch` has been completely rewritten and the syntax has changed.
  *  now "requires" a callback (technically there is an option to invoke it without one, but you have to go out of your way to do so -- read the docs)
  * match criteria must be specified as a function
    * `hs.axuielement.searchCriteriaFunction(matchCriteria)` can be used to use `hs.axuielement:matchesCriteria` for simple criteria functions
    * you can write your own more complex match functions (e.g. you could write a function that checks the role and if the role is an application, checks for one set of attributes but if the role is something else, checks a different set of attributes, etc.)

I think that's it for the methods that were originally documented... some of the undocumented ones have been removed... check the docs linked above for the full tally.

I'm going to close this, but reopen if you think I've missed anything.
